tests/unit/test_util.py

```diff
@@ -88,6 +93,33 @@ def test_git_commit_info(self):
         self.assertIsInstance(git_commit_info, str)
         self.assertEqual(git_commit_info[:2], "v2")
 
+    def test_have_optional_dependency(self):
+        with self.assertRaisesRegex(ModuleNotFoundError,"Optional dependency pybtex is not available. See https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html#optional-dependencies for more details."):
+            pybtex = sys.modules['pybtex']
+            sys.modules['pybtex'] = None
+            pybamm.print_citations()
+        with self.assertRaisesRegex(ModuleNotFoundError,"Optional dependency tqdm is not available. See https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html#optional-dependencies for more details."):
```

Comment:
These tests look good, just missing some coverage as Ferran mentions. Instead of hard-coding the name of the dependency, is it possible to mock the dependencies in one go and use f-strings? It might reduce some of the boilerplate in these assert statements.

Also, could you add a small sentence or two highlighting this to the Contributing guide? i.e., whenever one adds a dependency for optional functionality, they should write a corresponding unit test in `test_utilpy` to ensure that an error is raised upon the absence of said dependency.

Reply:
> Instead of hard-coding the name of the dependency, is it possible to mock the dependencies in one go and use f-strings?

I also thought about doing the same but because each dependency has distinct implementation in terms of functions & complexity(as we are not just checking the availability of a specific dependency but testing the dependent functions), I found myself unable to do that. We can try doing this afterwards too, happy to have your suggestions.

Reply:
Sounds good to me in that case. The error message has been stored as a variable, so checking each dependency is not a big deal, given that we don't have a lot of dependencies.